src/main/java/gregtech/api/items/toolitem/ItemGTToolbelt.java

```diff
@@ -0,0 +1,772 @@
+    public int getDefaultColor(ItemStack stack) {
+        return 0xA06540;
+    }
+
+    @Override
+    public boolean shouldGetContainerItem() {
+        return false;
+    }
+
+    public static boolean checkIngredientAgainstToolbelt(@NotNull ItemStack input, @NotNull OreIngredient ingredient) {
+        if (input.getItem() instanceof ItemGTToolbelt toolbelt) {
+            if (toolbelt.supportsIngredient(input, ingredient)) {
+                toolbelt.setOnCraftIngredient(input, ingredient);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static boolean checkToolAgainstToolbelt(@NotNull ItemStack toolbelt, @NotNull ItemStack tool) {
+        if (toolbelt.getItem() instanceof ItemGTToolbelt belt && ToolHelper.isUtilityItem(tool)) {
+            return belt.supportsTool(toolbelt, tool);
+        }
+        return false;
+    }
+
+    protected static class ToolbeltCapabilityProvider implements ICapabilityProvider, INBTSerializable<NBTTagCompound> {
+
+        protected final IntSupplier slotCountSupplier;
+
+        private @Nullable ToolStackHandler handler;
+
+        public ToolbeltCapabilityProvider(ItemStack stack) {
+            slotCountSupplier = () -> {
+                NBTTagCompound toolTag = stack.getOrCreateSubCompound(ToolHelper.TOOL_TAG_KEY);
+                String string = toolTag.getString(MATERIAL_KEY);
+                Material material = GregTechAPI.materialManager.getMaterial(string);
+                if (material == null) {
+                    toolTag.setString(MATERIAL_KEY, (material = Materials.Iron).getRegistryName());
+                }
+                ToolProperty toolProperty = material.getProperty(PropertyKey.TOOL);
+                return Math.min(8, 2 + (toolProperty == null ? 0 : toolProperty.getToolHarvestLevel()));
+            };
+        }
+
+        @Override
+        public boolean hasCapability(@NotNull Capability<?> capability, EnumFacing facing) {
+            return capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY;
+        }
+
+        @Override
+        public <T> T getCapability(@NotNull Capability<T> capability, EnumFacing facing) {
+            if (capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+                return CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.cast(this.getHandler(0));
+            else return null;
+        }
+
+        @Override
+        public NBTTagCompound serializeNBT() {
+            return this.getHandler(0).serializeNBT();
+        }
+
+        @Override
+        public void deserializeNBT(NBTTagCompound nbt) {
+            // make sure we can load all the slots, no matter what we're supposed to be limited to.
+            int minsize = nbt.hasKey("Size") ? nbt.getInteger("Size") : 0;
+            // .copy() prevents double damage ticks in singleplayer
+            this.getHandler(minsize).deserializeNBT(nbt.copy());
+        }
+
+        protected ToolStackHandler getHandler(int minsize) {
```

Comment:
I am kinda confused about the purpose of the `minsize` parameter here. ItemStackHandler will automatically fill an array of stacks based on the passed in size, so passing just `slotCountSupplier.getAsInt()` into ToolStackHandler should be all you need to do to create an inventory of a certain size, without needing to worry about slots.